interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -28,37 +28,355 @@
+
+        constraint_expr = ca.vertcat(*constraint_args)
+        upper_bound = ca.vertcat(
+            constraints.ubu,
+            constraints.ubx,
+            constraints.ug,
+            constraints.uh,
+            constraints.uphi
+        )
+        lower_bound = ca.vertcat(
+            constraints.lbu,
+            constraints.lbx,
+            constraints.lg,
+            constraints.lh,
+            constraints.lphi
+        )
+
+        constraint_expr_e = ca.vertcat(*constraint_args_e)
+        upper_bound_e = ca.vertcat(
+            constraints.ubx_e,
+            constraints.ug_e,
+            constraints.uh_e,
+            constraints.uphi_e
+        )
+        lower_bound_e = ca.vertcat(
+            constraints.lbx_e,
+            constraints.lg_e,
+            constraints.lh_e,
+            constraints.lphi_e
+        )
+
+        lower_violation = ca.fmax(lower_bound - constraint_expr, 0)
+        upper_violation = ca.fmax(constraint_expr - upper_bound, 0)
+
+        lower_violation_e = ca.fmax(lower_bound_e - constraint_expr_e, 0)
+        upper_violation_e = ca.fmax(constraint_expr_e - upper_bound_e, 0)
+
+        slack_indices = np.concatenate((
+            constraints.idxsbu,
+            constraints.idxsbx + dims.nbu,
+            constraints.idxsg + dims.nbu + dims.nbx,
+            constraints.idxsh + dims.nbu + dims.nbx + dims.ng,
+            constraints.idxsphi + dims.nbu + dims.nbx + dims.ng + dims.nh
+        ))
+
+        slack_indices_e = np.concatenate((
+            constraints.idxsbx_e,
+            constraints.idxsg_e  + dims.nbx_e,
+            constraints.idxsh_e + dims.nbx_e + dims.ng_e,
+            constraints.idxsphi_e + dims.nbx_e + dims.ng_e + dims.nh_e
+        ))
+
+        self.nonslacked_indices = np.setdiff1d(
+            np.arange(constraint_expr.shape[0]),
+            slack_indices
+        )
+
+        self.nonslacked_indices_e = np.setdiff1d(
+            np.arange(constraint_expr_e.shape[0]),
+            slack_indices_e
+        )
+
+        lower_slack_expression = lower_violation[slack_indices]
+        upper_slack_expression = upper_violation[slack_indices]
+
+        lower_slack_expression_e = lower_violation_e[slack_indices_e]
+        upper_slack_expression_e = upper_violation_e[slack_indices_e]
+
+        self.constraint_function = ca.Function(
+            'constraint_function',
+            cost_fun_args,
+            [lower_violation, upper_violation, lower_slack_expression, upper_slack_expression]
+        )
+
+        self.constraint_function_e = ca.Function(
+            'constraint_function_e',
+            cost_fun_args_e,
+            [lower_violation_e, upper_violation_e, lower_slack_expression_e, upper_slack_expression_e]
+        )
+
+    @property
+    def parameter_values(self):
+        """:math:`p` - initial values for parameter vector - can be updated stage-wise"""
+        return self.__parameter_values
+
+    @parameter_values.setter
```

Comment:
do we want to allow setting the parameters from outside? Or only via an update given a solver object?

Reply:
Removed it for now, as it may be confusing or used inappropriately